client/src/controllers/DrilldownController.test.js

```diff
@@ -0,0 +1,201 @@
+              <button id="action-element" data-controller="w-action">Action</button>
+            </section>`,
+    identifier = 'w-drilldown',
+  ) => {
+    document.body.innerHTML = `<main>${html}</main>`;
+
+    application = new Application();
+    application.handleError = (error, message) => {
+      errors.push({ error, message });
+    };
+    application.register(identifier, DrilldownController);
+    application.register('w-action', ActionController);
+    application.register('w-dropdown', DropdownController);
+    application.start();
+
+    await jest.runAllTimersAsync();
+
+    return [
+      ...document.querySelectorAll(`[data-controller~="${identifier}"]`),
+    ].map((element) =>
+      application.getControllerForElementAndIdentifier(element, identifier),
+    );
+  };
+
+  beforeEach(async () => {
+    await setup();
+  });
+
+  afterEach(() => {
+    application?.stop && application.stop();
+    errors = [];
+    eventNames.forEach((name) => {
+      events[name] = [];
+    });
+  });
+
+  describe('DrilldownController open/close actions', () => {
+    it('opens the submenu and sets active submenu value when toggle is clicked', async () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      const submenu = document.querySelector('#drilldown-field-0');
+
+      toggleButton.click();
+      await jest.runAllTimersAsync();
+
+      expect(toggleButton.getAttribute('aria-expanded')).toBe('true');
+      expect(submenu.hidden).toBeFalsy();
+    });
+
+    it('closes the submenu and clears active submenu value when close action is triggered', () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      toggleButton.click();
+
+      const closeButton = document.querySelector(
+        '[data-action="click->w-drilldown#close"]',
+      );
+      closeButton.click();
+
+      const submenu = document.querySelector('#drilldown-field-0');
+      expect(toggleButton.getAttribute('aria-expanded')).toBe('false');
+      expect(submenu.hidden).toBeTruthy();
+    });
+  });
+
+  describe('DrilldownController delayedClose method', () => {
+    it('should call close after 200ms delay', () => {
+      document.body.innerHTML = `
+      <div id="drilldown-field-0">
+        <button data-action="w-dropdown:hide->w-drilldown#delayedClose">Close</button>
+      </div>`;
+
+      const submenu = document.querySelector('#drilldown-field-0');
+      const delayCloseButton = document.querySelector(
+        '[data-action="w-dropdown:hide->w-drilldown#delayedClose"]',
+      );
+      submenu.setAttribute('data-w-drilldown-active', 'true');
+
+      expect(delayCloseButton).not.toBeNull();
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('true');
+
+      delayCloseButton.click();
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('true');
+
+      jest.advanceTimersByTime(200);
```

Comment:
You may need to use an async version of this, because when attributes update, this is because some promises need to resolve.

You will first need to ensure this test supports async.

```
    it('should call close after 200ms delay', async () => {
```


Then use the async version of the timer.

```suggestion
      await jest.advanceTimersByTimeAsync(200);
```

Reply:
I did this, but the test still did not pass.
![image](https://github.com/user-attachments/assets/a5b52473-2208-452c-a879-a222a187640a)

Reply:
I have become tired of trying to check why it is failing. I found that this test is not actually calling the `delayedClose() `method because it's not printing the message I have written in the method to check if it's calling the delayed method or not. But I can't find the reason yet. I used `data-action="w-dropdown:hidden->w-drilldown#delayedClose"` as suggested in the comment of the method, but it's still not working. As you can see in the attachment also-

![image](https://github.com/user-attachments/assets/3b10dc4b-97ba-48e6-9fff-8c87d2526d37)
![image](https://github.com/user-attachments/assets/f83f48f7-549a-4ee0-a4c2-669f5740605d)
![image](https://github.com/user-attachments/assets/8d3c66c5-dd7c-4bd5-8dd0-f4ae77d72c56)

Reply:
Now , I think,I should throw my laptop and sleep....only this can fix me now 🥲

Reply:
I feel like this also sometimes. I'll try to take a look soon and see if I can help.